File: packages/plugin/internal/rename/transcript.go
package rename

import (
	"bufio"
	"encoding/json"
	"os"
	"strings"
)

var secretPatterns = []*regexpWrapper{
	mustRegexp(`(?i)(api[_-]?key|access[_-]?token|secret|password|passwd|authorization)\s*[:=]\s*["']?[^"'\s]+`),
	mustRegexp(`sk-[A-Za-z0-9_-]{20,}`),
	mustRegexp(`ghp_[A-Za-z0-9_]{20,}`),
}

func ReadTranscriptTail(path string, maxBytes int64) TranscriptTail {
	if path == "" {
		return TranscriptTail{}
	}
	if maxBytes <= 0 {
		maxBytes = 384 * 1024
	}
	file, err := os.Open(path)
	if err != nil {
		return TranscriptTail{}
	}
	defer file.Close()
	stat, err := file.Stat()
	if err != nil {
		return TranscriptTail{}
	}
	start := stat.Size() - maxBytes
	if start < 0 {
		start = 0
	}
	if _, err := file.Seek(start, 0); err != nil {
		return TranscriptTail{}
	}

	userMessages := []string{}
	assistantMessages := []string{}
	toolSignals := []string{}
	scanner := bufio.NewScanner(file)
	scanner.Buffer(make([]byte, 64*1024), 2*1024*1024)
	for scanner.Scan() {
		line := scanner.Bytes()
		if len(line) == 0 {
			continue
		}
		var parsed map[string]any
		if err := json.Unmarshal(line, &parsed); err != nil {
			continue
		}
		payload, ok := parsed["payload"].(map[string]any)
		if !ok {
			continue
		}
		collectFromPayload(payload, &userMessages, &assistantMessages, &toolSignals)
	}
	return TranscriptTail{
		UserMessages:      tail(userMessages, 3),
		AssistantMessages: tail(assistantMessages, 2),
		ToolSignals:       tail(toolSignals, 12),
	}
}

func collectFromPayload(payload map[string]any, userMessages, assistantMessages, toolSignals *[]string) {
	switch payload["type"] {
	case "message":
		text := extractTextContent(payload["content"])
		if text == "" {
			return
		}
		switch payload["role"] {
		case "user":
			*userMessages = append(*userMessages, text)
		case "assistant":
			*assistantMessages = append(*assistantMessages, text)
		}
	case "response_item":
		if nested, ok := payload["payload"].(map[string]any); ok {
			collectFromPayload(nested, userMessages, assistantMessages, toolSignals)
		}
	case "event_msg":
		if nested, ok := payload["payload"].(map[string]any); ok && nested["type"] == "task_complete" {
			*toolSignals = append(*toolSignals, "task completed")
		}
	case "function_call", "tool_call":
		name, ok := payload["name"].(string)
		if !ok || name == "" {
			name = "tool"
		}
		*toolSignals = append(*toolSignals, name)
	}
}

func extractTextContent(content any) string {
	switch value := content.(type) {
	case string:
		return normalizeTranscriptText(value)
	case []any:
		parts := []string{}
		for _, item := range value {
			record, ok := item.(map[string]any)
			if !ok {
				continue
			}
			for _, key := range []string{"text", "input_text", "output_text"} {
				if text, ok := record[key].(string); ok {
					parts = append(parts, text)
					break
				}
			}
		}
		return normalizeTranscriptText(strings.Join(parts, " "))
	default:
		return ""
	}
}

func normalizeTranscriptText(input string) string {
	return clip(normalizeWhitespace(redactSecrets(input)), 2500)
}

func redactSecrets(input string) string {
	output := input
	for _, pattern := range secretPatterns {
		output = pattern.ReplaceAllString(output, "[REDACTED]")
	}
	return output
}

func tail(values []string, count int) []string {
	if len(values) <= count {
		return values
	}
	return values[len(values)-count:]
}

File: .old/src/transcript.ts
import { readFileSync, statSync } from 'node:fs'

import { clip, normalizeWhitespace, redactSecrets } from './text.ts'

export interface TranscriptTail {
	userMessages: string[]
	assistantMessages: string[]
	toolSignals: string[]
}

export function readTranscriptTail(
	path: string | null | undefined,
	maxBytes = 384 * 1024
): TranscriptTail {
	if (!path) {
		return { userMessages: [], assistantMessages: [], toolSignals: [] }
	}
	try {
		const stat = statSync(path)
		const start = Math.max(0, stat.size - maxBytes)
		const raw = readFileSync(path).subarray(start).toString('utf8')
		const lines = raw.split('\n').filter(Boolean)
		const userMessages: string[] = []
		const assistantMessages: string[] = []
		const toolSignals: string[] = []
		for (const line of lines) {
			try {
				const parsed = JSON.parse(line) as Record<string, unknown>
				const payload = parsed.payload as Record<string, unknown> | undefined
				if (!payload) {
					continue
				}
				collectFromPayload(payload, userMessages, assistantMessages, toolSignals)
			} catch {
				// Transcript JSONL can be truncated at the head; ignore partial lines.
			}
		}
		return {
			userMessages: userMessages.slice(-3),
			assistantMessages: assistantMessages.slice(-2),
			toolSignals: toolSignals.slice(-12)
		}
	} catch {
		return { userMessages: [], assistantMessages: [], toolSignals: [] }
	}
}

function collectFromPayload(
	payload: Record<string, unknown>,
	userMessages: string[],
	assistantMessages: string[],
	toolSignals: string[]
): void {
	if (payload.type === 'message') {
		const role = payload.role
		const text = extractTextContent(payload.content)
		if (text) {
			if (role === 'user') {
				userMessages.push(text)
			} else if (role === 'assistant') {
				assistantMessages.push(text)
			}
		}
		return
	}
	if (payload.type === 'response_item') {
		const nested = payload.payload as Record<string, unknown> | undefined
		if (nested) {
			collectFromPayload(nested, userMessages, assistantMessages, toolSignals)
		}
		return
	}
	if (payload.type === 'event_msg') {
		const nested = payload.payload as Record<string, unknown> | undefined
		if (nested?.type === 'task_complete') {
			toolSignals.push('task completed')
		}
		return
	}
	if (payload.type === 'function_call' || payload.type === 'tool_call') {
		const name = typeof payload.name === 'string' ? payload.name : 'tool'
		toolSignals.push(name)
	}
}

function extractTextContent(content: unknown): string {
	if (typeof content === 'string') {
		return normalizeTranscriptText(content)
	}
	if (!Array.isArray(content)) {
		return ''
	}
	const parts: string[] = []
	for (const item of content) {
		if (!item || typeof item !== 'object') {
			continue
		}
		const record = item as Record<string, unknown>
		const text = record.text ?? record.input_text ?? record.output_text
		if (typeof text === 'string') {
			parts.push(text)
		}
	}
	return normalizeTranscriptText(parts.join(' '))
}

function normalizeTranscriptText(input: string): string {
	return clip(normalizeWhitespace(redactSecrets(input)), 2500)
}
